### PR TITLE
Support env variables for config (#111)

### PR DESCRIPTION
**Pull request by nikonhub**
closes : https://github.com/Boeing/config-file-validator/issues/111

We could have used os.Getenv() directly as default value for flag.String(). But the are multiple uses of isFlagSet() in the code, which would return false even if the variable is set through the env.

### CODE REVIEW

**Comment by kehoecj** [comment]
@nikonhub Thanks for the PR! Approach looks solid - please address the golangci-lint findings

File: cmd/validator/validator.go [hunk 1]
```diff
@@ -161,6 +170,15 @@ func isFlagSet(flagName string) bool {
...
 }
 
+func setFlagFromEnvIfNotSet(flagName string, envVar string) {
+	if !isFlagSet(flagName) {
+		envVarValue, isEnvVarSet := os.LookupEnv(envVar)
+		if isEnvVarSet {
+			flag.Set(flagName, envVarValue)
+		}
+	}
+}
```
Comment (ccoVeille) [inline cmd/validator/validator.go:180]: I would avoid a function wide `if`, and use early return

```suggestion
func setFlagFromEnvIfNotSet(flagName string, envVar string) {
	if isFlagSet(flagName) {
		return
	}
	envVarValue, isEnvVarSet := os.LookupEnv(envVar)
	if isEnvVarSet {
		flag.Set(flagName, envVarValue)
	}
}
```

**Comment by nikonhub** [comment]
@kehoecj There is a gocyclo error. But I also see another PR opened with the same error on the same function. How should we handle this ?

File: cmd/validator/validator.go [hunk 2]
```diff
@@ -80,6 +80,23 @@ func getFlags() (validatorConfig, error) {
 	versionPtr := flag.Bool("version", false, "Version prints the release version of validator")
 	groupOutputPtr := flag.String("groupby", "", "Group output by filetype, directory, pass-fail. Supported for Standard and JSON reports")
 	quietPrt := flag.Bool("quiet", false, "If quiet flag is set. It doesn't print any output to stdout.")
```
Comment (ccoVeille) [inline cmd/validator/validator.go:82]: Not related to your changes, but I think there is a typo here


```suggestion
 	quietPtr := flag.Bool("quiet", false, "If quiet flag is set. It doesn't print any output to stdout.")
```

File: cmd/validator/validator.go [hunk 3]
```diff
@@ -79,7 +79,24 @@ func getFlags() (validatorConfig, error) {
...
+	quietPtr := flag.Bool("quiet", false, "If quiet flag is set. It doesn't print any output to stdout.")
+
+	flagsEnvMap := map[string]string{
+		"depth":              "CSV_DEPTH",
+		"exclude-dirs":       "CSV_EXCLUDE_DIRS",
+		"exclude-file-types": "CSV_EXCLUDE_FILE_TYPES",
+		"output":             "CSV_OUTPUT",
+		"reporter":           "CSV_REPORTER",
+		"groupby":            "CSV_GROUPBY",
+		"quiet":              "CSV_QUIET",
```
Comment (kehoecj) [inline cmd/validator/validator.go:91]: Why the `CSV` prefix? Did you mean `CFV`?

Comment (nikonhub) [reply cmd/validator/validator.go]: It was the habit to type "CSV". Sorry about that. I fixed it.

**Comment by kehoecj** [comment]
> @kehoecj There is a gocyclo error. But I also see another PR opened with the same error on the same function. How should we handle this ?

Ignore - after hacktoberfest I'll go back and do some refactoring to satisfy gocyclone

**Review by kehoecj — CHANGES_REQUESTED** [review]
Please add a section to the README.md describing how users can use environment variables in addition to or instead of the command line arguments.

Make sure it's clear that a command line var will override an env var

File: README.md [hunk 4]
```diff
@@ -116,6 +116,18 @@ optional flags:
...
+
+The config-file-validator supports setting options via environment variables. If both command-line flags and environment variables are set, the command-line flags will take precedence. The supported environment variables are as follows:
+
+- `CFV_DEPTH`: Equivalent to the `-depth` flag
+- `CFV_EXCLUDE_DIRS`: Equivalent to the `-exclude-dirs` flag
+- `CFV_EXCLUDE_FILE_TYPES`: Equivalent to the `-exclude-file-types` flag
+- `CFV_OUTPUT`: Equivalent to the `-output` flag
+- `CFV_REPORTER`: Equivalent to the `-reporter` flag
+- `CFV_GROUPBY`: Equivalent to the `-groupby` flag
+- `CFV_QUIET`: Equivalent to the `-quiet` flag
```
Comment (kehoecj) [inline README.md:129]: Would prefer these in a markdown table - easier to read/maintain

Comment (nikonhub) [reply README.md]: 👍 Just changed it

**Review by kehoecj — APPROVED** [review]
LGTM - thanks for the PR @nikonhub !

**Pull request merged**